### PR TITLE
5.0.4: Fix errors on data deletion and project load

### PR DESCRIPTION
**Pull request by krzywon**
This corrects issues resulting in errors thrown when data is deleted or new projects are loaded noted by @smk78 in #1772 points [6](https://github.com/SasView/sasview/issues/1772#issuecomment-768257509) and [13](https://github.com/SasView/sasview/issues/1772#issuecomment-768422908) and by @RichardHeenan in multiple comments of #1689.

### CODE REVIEW

**Comment by butlerpd** [comment]
ready for testing on Win

**Comment by wpotrzebowski** [comment]
ready for testing on OSX

File: src/sas/qtgui/Plotting/QRangeSlider.py [hunk 1]
```diff
@@ -230,6 +241,10 @@ def onRelease(self, ev):
 
     def clear_markers(self):
         """
-        Should be no way to clear the markers
+        Disconnect the input and clear the callbacks
         """
-        pass
+        if self.input:
+            self.input.textChanged.disconnect(self.inputChanged)
```
Comment (rozyczko) [inline src/sas/qtgui/Plotting/QRangeSlider.py:247]: This will get disconnected when the slider is closed/removed but explicit disconnect is always good. We should probably strive at having disconnects for each signal connect.

**Review by rozyczko — APPROVED** [review]
Good code, very important fixes which should be in the code since the beginning. Thanks for fixing!

**Comment by wpotrzebowski** [comment]
Tested several scenarios with project saving/loading and deleting files and it seems to work fine. 
However I run into error when I try open project saving dialog and click on Cancel:
```
20:12:16 - ERROR: Traceback (most recent call last):
  File "site-packages/sasview-5.0.4-py3.6.egg/sas/qtgui/MainWindow/GuiManager.py", line 693, in actionSave_Project
TypeError: expected str, bytes or os.PathLike object, not NoneType 
```
I suppose this may be some other issue though and program seems to recover from it (doesn't crash).

**Comment by krzywon** [comment]
I made a small change to address the cancel issue as well. Should I create new win and OSX builds to test?

**Comment by wpotrzebowski** [comment]
ready for testing on OSX

**Comment by butlerpd** [comment]
I am finding some serious functionality issues. I have so far only tested the q sliders (using windows installer) as they were a large part of the code touched it seems. The main problem so far seems to be with the P(r) perspective.  In this case I find:
* the sliders may initially be dragable graphically (the point of having them) or not. I have yet to track down precisely what triggers them to become dragable. However once dragged (see next point) and calculate pressed they are no longer dragable. 
* Once they are dragable (the line highlights in red) and one starts to drag, it never lets go. i.e. it keeps moving with the mouse no matter how much one clicks or moves off the graph and back or clicks on another graph and comes back. I have found that clicking on compute again stops it cold ... but they are no longer dragable (see above).
* Finally, and most problematic is editing the min or max box directly.  Here the problem seems to be linked to number that evaluates to zero.  So if qmin is 0.00154 and I backspace one a time I get 0.0015, 0.001, **BLANK** typing 0 remains blank. typing 0. yields . and 0.0 goes back to blank. typing 0.01 will often yield 0.01xx. I do not know where xx came from. The error thrown in the console every time is:
 ```
14:01:48 - ERROR: Traceback (most recent call last):
  File "site-packages\sasview-5.0.4-py3.6.egg\sas\qtgui\Plotting\QRangeSlider.py", line 184, in inputChanged
  File "site-packages\sasview-5.0.4-py3.6.egg\sas\qtgui\Plotting\QRangeSlider.py", line 177, in _get_q
ValueError: could not convert string to float:
```
* However changing the editable box (when it works correctly)  **does** change the slider positions on the graph.

In the case of the Invariant for which the sliders are not active, the qmin and qmax are not editable (is this a known problem or new?)

Finally, and strangely to me, the sliders and max min editable boxes work perfectly as before in the case of both Corfunc and Fitting.

**Comment by wpotrzebowski** [comment]
I can confirm that error on "Cancel" has been now fixed. (Tested on OSX).

**Comment by butlerpd** [comment]
Regarding @RichardHeenan in #1689, the QLineEdit errors **continue** to be thrown when saving the project with a single model and no data.  The same error mentioned by @smk78 in point [13](https://github.com/SasView/sasview/issues/1772#issuecomment-768422908). However the errors are no longer thrown when loading those projects back into a fresh instance of SasView.  However However in the case of the model only project of @RichardHeenan the fitpage is empty (no model is chosen) and the theory box in dataexplorer has one element titled M1 instead of M1 [data].

**Comment by butlerpd** [comment]
deleted data do not cause errors any more

**Comment by butlerpd** [comment]
Further research suggests that the edit box problem is related to the issue of the editor validating/calculating after each key stroke.  In fact it seems to maybe be slightly better than originally. One problem is that the value **must** be between qmin and qmax of the data. So before this PR, of the data qmin is 0.0055 and that is the default in the min q box, as I delete the first 5 (to replace say with 0.008) the value would effectively be 0.005 momentarily which is less than qmin and thus not allowed so nothing would happen.  Now it will delete but deleting the second makes it go blank. Further inserting a 1, 2, or 3 in front of 0.0055 will make the number jump back to 0.0055 (this may be OK even if a bit enerving for those not in the know).  On the other hand inserting an 8 in front of 0.0055 does so but jumps the cursor back to the end of 0.00855.  This is just very enerving.  I will need to see how far back this goes or if it is new with the sliders.

**Comment by wpotrzebowski** [comment]
ready for testing on OSX

**Comment by krzywon** [comment]
I fixed this by enabling the interactive features of the plots after clearing data for replacement.
> - the sliders may initially be dragable graphically (the point of having them) or not. I have yet to track down precisely what triggers them to become dragable. However once dragged (see next point) and calculate pressed they are no longer dragable.

This was related to reinititializing the q-slider values in PlotterData to fix the save/load errors. I moved this reset into the save project routines and things seem better now.
> - Once they are dragable (the line highlights in red) and one starts to drag, it never lets go. i.e. it keeps moving with the mouse no matter how much one clicks or moves off the graph and back or clicks on another graph and comes back. I have found that clicking on compute again stops it cold ... but they are no longer dragable (see above).

I changed the box to update when editing is finished, not when text is changed. This should be more stable now.
> - Finally, and most problematic is editing the min or max box directly. Here the problem seems to be linked to number that evaluates to zero. So if qmin is 0.00154 and I backspace one a time I get 0.0015, 0.001, BLANK typing 0 remains blank. typing 0. yields . and 0.0 goes back to blank. typing 0.01 will often yield 0.01xx. I do not know where xx came from. The error thrown in the console every time is:

This was not present running the code locally, but I tested the last Windows build and also saw this error suggesting the merge might revert something. I am hoping my recent changes fixed this, and I will generate a new windows installer to check. Otherwise, I will need to merge the release_5.0.4 branch and fix what the merge is breaking.
> Regarding @RichardHeenan in #1689, the QLineEdit errors continue to be thrown when saving the project with a single model and no data. The same error mentioned by @smk78 in point 13.

**Comment by krzywon** [comment]
ready for testing on Win

**Comment by krzywon** [comment]
ready for testing on OSX

**Review by butlerpd — APPROVED** [review]
New code was not reviewed but functionality was tested with the installer build on windows and seems ready for merging. Anything else should be moved to other issues/PR.  In particular:
1. fix to error being thrown when using the Edit Mask menu item is fixed.  I still think the final/proper fix should disable the menu item just like others are disabled when not usable and as the button that has the same function is disabled.  However, not only is the error not thrown a more useful message is added to the logger and a popup explains to the user why it does not work and what they need to do. **nice**
2. Inversion sliders work very nicely and smoothly now - seems even better than originally. Of course all sliders have a bit of sluggish feel to them but that detail would be a future major project I think.  Very nice
3. Inversion q limits editor also work very smoothly now. The only minor niggle is that hitting return (or clicking away) does not update the slider position as it does in the corfunc perspective for example. However as soon as you hit calculate the sliders move appropriately.  Plenty good enough for this release. Eventually it would be nice for all perspectives to behave similarly.
4. Fitting perspective.  the sliders seem to be working much more smoothly now as well -- in fact look just like they do with the P(r) - **I like it!**
5. The QLineEdit errors on saving a project seems to be fixed. At least I can no longer make it happen.
6.  The problem described by @RichardHeenan on loading a project with only a theory model remains.  That should however I think be moved to another issue/ticket as I don't think it is related to this PR
7. As discussed by @smk78 the plot state of child data is not respected on loading a project but they are loaded in the data manager.  This is probably not related to this PR and IMO should be submitted as a separate ticket.
8. A new observation but again probably should not be part of this PR. Load a data file and send to inversion. set the appropriate parameters and press `calculate` then save as project. Load a fresh sasview and load the project. Everything loads correctly, including the slider positions ... almost.  The number of terms does not seem to be respected. Moreover hitting calculate to display the plots leads to a P(r) plot that is a damped ringing centered at zero no matter what number of terms is chosen. Only by changing the Dmax ever so slightly and calculation will the computation return something sensible. It is really baffling what calculation is being reported instead of the P(r) and why. strange what computation it may be doing instead of P(r). One other observation is that the I_{fit} and P_{fit} subplots are loaded correctly as evidenced by sending them to Plot before recalculating anything.

Bottom line I strongly recommend merging with the caveat that the remaining items above be properly documented in new issues.
